servant-server/src/Servant/Server/Internal.hs

```diff
@@ -307,139 +238,82 @@ instance (KnownSymbol sym, FromText a, HasServer sublayout)
--- was @*\/\*@, we return encode using the first @Content-Type@ type on the
--- list.
+-- If successfully returning a value, we use the type-level list,
+-- combined with the request's @Accept@ header, to encode the value
+-- for you (returning the default status code for the method -
+-- e.g. for GET the status code will be 200). If there was no @Accept@
+-- header or it was @*\/\*@, we encode the result using the first
+-- @Content-Type@ type in the list.
 instance
 #if MIN_VERSION_base(4,8,0)
          {-# OVERLAPPABLE #-}
 #endif
          ( AllCTRender ctypes a
-         ) => HasServer (Post ctypes a) where
-
-  type ServerT (Post ctypes a) m = m a
+         , KnownSymbol method
+         , KnownNat (DefaultStatusCode method)
+         ) => HasServer (HttpMethod method ctypes a) where
 
-  route Proxy = methodRouter methodPost (Proxy :: Proxy ctypes) created201
-
-instance
-#if MIN_VERSION_base(4,8,0)
-         {-# OVERLAPPING #-}
-#endif
-         HasServer (Post ctypes ()) where
+  type ServerT (HttpMethod method ctypes a) m = m a
 
-  type ServerT (Post ctypes ()) m = m ()
+  route Proxy = methodRouter method (Proxy :: Proxy ctypes) status
+    where 
+      method = B8.pack $ symbolVal (Proxy :: Proxy method)
+      status = toEnum $ fromInteger $ natVal (Proxy :: Proxy (DefaultStatusCode method))
 
-  route Proxy = methodRouterEmpty methodPost
 
--- Add response headers
 instance
 #if MIN_VERSION_base(4,8,0)
          {-# OVERLAPPING #-}
-#endif
-         ( GetHeaders (Headers h v), AllCTRender ctypes v
-         ) => HasServer (Post ctypes (Headers h v)) where
-
-  type ServerT (Post ctypes (Headers h v)) m = m (Headers h v)
-
-  route Proxy = methodRouterHeaders methodPost (Proxy :: Proxy ctypes) created201
-
--- | When implementing the handler for a 'Put' endpoint,
--- just like for 'Servant.API.Delete.Delete', 'Servant.API.Get.Get'
--- and 'Servant.API.Post.Post', the handler code runs in the
--- @EitherT ServantErr IO@ monad, where the 'Int' represents
--- the status code and the 'String' a message, returned in case of
--- failure. You can quite handily use 'Control.Monad.Trans.EitherT.left'
--- to quickly fail if some conditions are not met.
---
--- If successfully returning a value, we use the type-level list, combined
--- with the request's @Accept@ header, to encode the value for you
--- (returning a status code of 200). If there was no @Accept@ header or it
--- was @*\/\*@, we return encode using the first @Content-Type@ type on the
--- list.
-instance
-#if MIN_VERSION_base(4,8,0)
-         {-# OVERLAPPABLE #-}
-#endif
-         ( AllCTRender ctypes a) => HasServer (Put ctypes a) where
+#endif 
+        ( AllCTRender ctypes a
+        , KnownSymbol method
+        ) => HasServer (HttpMethod method ctypes (ServantSuccess a)) where
 
-  type ServerT (Put ctypes a) m = m a
+  type ServerT (HttpMethod method ctypes (ServantSuccess a)) m = m (ServantSuccess a)
```

Comment:
Hm, thinking more on it, I think we want to go in the direction of making HTTP status codes more rather than less knowable statically. (And the handlers less rather than more HTTP-aware.) Maybe the status code can be a (Nat-kinded) param to `HttpMethod`?